test/main.cpp

```diff
@@ -215,6 +215,22 @@ static bool test13()
     }
 }
 
+static bool test14()
+{
+    // Avoid cumulated error
+    PolylineEncoder encoder;
+    encoder.addPoint(0.0000005, 0.0000005);
+    encoder.addPoint(0.0000000, 0.0000000);
+
+    // Intentionally not use test() function as the precision cut generates difference between encode and decode
+    if (encoder.encode() == "????") {
```

Comment:
I am just curious, why do you think that `"????"` is the correct result of encoding?

Reply:
because this site: https://developers.google.com/maps/documentation/utilities/polylineutility
gives it as result. 

and as the encoding algorithm truncates the geocoordinates, then the trailing part of the points (part after e5) should not affect the result of encoding.

Reply:
Are you able to encode the same points `(0.0000005, 0.0000005)` and `(0.0, 0.0)` with that site? When I try to use `0.0000005` as a coordinate (which indeed pretty small value and, IMO, unrealistic to be a coordinate) it creates two points: `(0.0, 0.0)` and `(0.0, 0.0)`. For two zero points it's, indeed` gives `????`, but in your test here your don't enter two identical points. The algorithm doesn't treat them equal as well, but you changed that behavior, which I don't agree with.

Reply:
> (which indeed pretty small value and, IMO, unrealistic to be a coordinate)

it's absolutely not unrealistic, real life scenario brought me to this PR :)

> it creates two identical points: (0.0, 0.0) and (0.0, 0.0).

And after my change your lib does the same! now it works like google intended!

The https://developers.google.com/maps/documentation/utilities/polylinealgorithm site's very first sentence states that it is a lossy algorithm (and by the way, this means, that your test function fails on valid inputs, since it does not consider about losing precision).

> The encoding format within a polyline needs to represent two coordinates representing latitude and longitude to a reasonable precision.

This sentence in the paper means to me, that the algorithm should handle the truncation


And my main reason: right now the lib gives a faulty result; in a real-life scenario when i have a bounch of geocoordinates from a hi-res map content and i encode it, and try to visualize it on google maps, the encoded format deviates from the real path.

Reply:
Well, you brought up valid points and I understand your concerns. I got it. But can we than simplify the patch. For example, this one should have the same effect:
````
diff --git a/src/polylineencoder.cpp b/src/polylineencoder.cpp
index e53c983..f97f59b 100644
--- a/src/polylineencoder.cpp
+++ b/src/polylineencoder.cpp
@@ -54,7 +54,7 @@ std::string PolylineEncoder::encode(double value)

     e5 <<= 1;                                     // (4)

-    if (value < 0) {
+    if (e5 != 0 && value < 0) {
         e5 = ~e5;                                 // (5)
     }
````

Reply:
I do not agree with you because of two reasons:

1) i'm not pretty sure that this is sufficent, but i'll try it out
2) since the encoding is a differentiation algorithm, logically the previous point should be a truncated one as well

Reply:
The thing is that `PolylineEncoder::encode(double)` processes the **difference** between coordinates. If it's less than the precision it's truncated anyways, but the problem comes from the negative differences (`value < 0`). For example, if you swap the points in your test14 you will get the expected result: `????` even without patching the algorithm.

Thus, could you please try the proposed patch, as I tend to preserve the current implementation and minimize the changes.

Reply:
i tried out your diff. for the record: my first fix was the very same, it handles weird cases.
but with just that the encoding is still not perfect, there is still deviation, but with my fix there is no. however i didnt find the reason yet, hence i didnt write testcase for that yet.

Reply:
I see. Can you please add another test that fails with the proposed (simple) patch?

Reply:
yeah, i'm on it

Reply:
the next commit has a real-life scenario in it where the minimal fix fails